### PR TITLE
Also export/import table tl_files in theme export/import

### PR DESCRIPTION
**Pull request by ausi**
As discussed in #7449

### CODE REVIEW

File: system/modules/core/classes/Theme.php [hunk 1]
```diff
@@ -548,8 +546,8 @@ protected function extractThemeFiles($arrFiles, $arrDbFields)
 							}
 						}
 
-						// Adjust the file paths in style sheets
-						elseif (\Config::get('uploadPath') != 'files' && ($table == 'tl_style_sheet' || $table == 'tl_style') && strpos($value, 'files') !== false)
+						// Adjust the file paths in style sheets and tl_files
+						elseif (\Config::get('uploadPath') != 'files' && ($table == 'tl_style_sheet' || $table == 'tl_style' || ($table == 'tl_files' && $name == 'path')) && strpos($value, 'files') !== false)
```
Comment (leofeyer) [inline system/modules/core/classes/Theme.php]: Why did you add an additional check for the field name here?

```
($table == 'tl_files' && $name == 'path'))
```

Are there other fields which potentially contain `files` and shall not be changed?

File: system/modules/core/classes/Theme.php [hunk 2]
```diff
@@ -647,8 +645,16 @@ protected function extractThemeFiles($arrFiles, $arrDbFields)
 						new \Folder($set['templates']);
 					}
 
-					// Update the datatbase
-					$this->Database->prepare("INSERT INTO ". $table ." %s")->set($set)->execute();
+					if ($table == 'tl_files')
+					{
+						// Update the database
+						$this->Database->prepare("UPDATE ". $table ." %s WHERE path = ?")->set($set)->execute($set['path']);
```
Comment (leofeyer) [inline system/modules/core/classes/Theme.php]: Why do you update instead of insert?

File: system/modules/core/classes/Theme.php [hunk 3]
```diff
@@ -958,14 +989,14 @@ protected function addTableTlImageSize(\DOMDocument $xml, \DOMElement $tables, \
...
-	protected function addDataRow(\DOMDocument $xml, \DOMElement $table, \Database\Result $objData, array $arrOrder=array())
+	protected function addDataRow(\DOMDocument $xml, \DOMElement $table, array $arrRow, array $arrOrder=array())
 	{
 		$t = $table->getAttribute('name');
 
 		$row = $xml->createElement('row');
 		$row = $table->appendChild($row);
 
-		foreach ($objData->row() as $k=>$v)
+		foreach ($arrRow as $k=>$v)
```
Comment (leofeyer) [inline system/modules/core/classes/Theme.php]: Why is this change required?

File: system/modules/core/classes/Theme.php (hunk 1 above)
Comment (ausi) [inline system/modules/core/classes/Theme.php]: An extension may add additional columns to `tl_files` which we shoudn't modify IMO.

File: system/modules/core/classes/Theme.php (hunk 2 above)
Comment (ausi) [inline system/modules/core/classes/Theme.php]: Because the entries were already inserted by the DBAFS

File: system/modules/core/classes/Theme.php (hunk 3 above)
Comment (ausi) [inline system/modules/core/classes/Theme.php]: To be able to call this method from `addFolderToArchive()` at line 1139

File: system/modules/core/classes/Theme.php (hunk 1 above)
Comment (leofeyer) [inline system/modules/core/classes/Theme.php]: Really? We shall only fix our own fields and leave third-party developers with the wrong paths?

Comment (ausi) [inline system/modules/core/classes/Theme.php]: IMO we should only change fields we know which data is in it. Just because a column value starts with `files/` doesn't ensure that it's a path that should be changed.

Comment (ausi) [inline system/modules/core/classes/Theme.php]: A third-party developer should use `$GLOBALS['TL_HOOKS']['extractThemeFiles']` to implement such transformations.

**Comment by leofeyer** [comment]
There is a big problem: You cannot use the `FilesModel` if the table is locked. See here:

https://github.com/contao/core/blob/c6ce781564307040655a85fb48834c8e615e41ed/system/modules/core/classes/Theme.php#L578

You have to use the `Database` class instead. And then you won't have to change line 999 either.

**Comment by ausi** [comment]
Aren’t they sharing the same database object? Why are the locked tables a problem?

**Comment by leofeyer** [comment]
I don't remember :( I only stumbled upon my comment, therefore I figure there must be an issue.

**Comment by ausi** [comment]
The comment was added in 8b71567fef283d7d7630538918474176ba71d9b1, but i don’t see a reason for it.

Is this related to #7345?

Maybe we should remove `if ($table != 'tl_files')` in [system/modules/core/classes/Theme.php:385](https://github.com/contao/core/blob/f8f5d3c72e0a4eb84f6367ed08d19d50d9ead483/system/modules/core/classes/Theme.php#L385).

**Comment by ausi** [comment]
I just took a closer look. Where did I use the `FilesModel` class while tables are locked? I only found occurrences of it in the export methods which don’t lock the tables.

**Comment by leofeyer** [comment]
Here's the adjusted version: https://github.com/contao/core/compare/develop...feature/export-file-paths

@ausi Please review and check whether it works for you.

**Comment by ausi** [comment]
I still don’t like converting all columns of `tl_files` with `customizeUploadPath()`. If you have a column in `tl_files` containing a serialized array like `a:1:{s:1:"a";a:1:{s:1:"b";a:1:{s:1:"c";s:13:"foo files bar";}}}` it would be converted to `a:1:{s:1:"a";a:1:{s:1:"b";s:5:"Array";}}` which is bad. E.g. `tl_files.meta` would also be broken if the array was nested one level deeper.

Your implementation is a bit different than mine: If `tl_files` is out of sync, you export nonexistent file entries, old hash values and don’t export file entries for files which are missing in the database. IMO it would be great to be consistent with the files we are exporting, that’s why I wrote it that way.

**Comment by leofeyer** [comment]
> If you have a column in `tl_files` containing a serialized array

We have this problem in the other tables, too. But it is impossible to foresee which columns people add, therefore we either have to fix all columns or agree on exporting only a defined set of fields (e.g. `path`, `meta` and the important part fields).

> is out of sync, you export nonexistent file entries, old hash values and don’t export file entries for files which are missing in the database

We should rather add a warning to sync the file system instead of trying to work around this IMHO.

**Comment by ausi** [comment]
> > If you have a column in `tl_files` containing a serialized array

> We have this problem in the other tables, too. But it is impossible to foresee which columns people add, therefore we either have to fix all columns or agree on exporting only a defined set of fields (e.g. `path`, `meta` and the important part fields).

I don’t get it. If someone uses `inputType`s that need conversion we already convert the values for them in [Theme.php:353](https://github.com/contao/core/blob/336c6a4e2ace2c878a797d23211140588b72b832/system/modules/core/classes/Theme.php#L535), [553](https://github.com/contao/core/blob/336c6a4e2ace2c878a797d23211140588b72b832/system/modules/core/classes/Theme.php#L553), [574](https://github.com/contao/core/blob/336c6a4e2ace2c878a797d23211140588b72b832/system/modules/core/classes/Theme.php#L574), [989](https://github.com/contao/core/blob/336c6a4e2ace2c878a797d23211140588b72b832/system/modules/core/classes/Theme.php#L989) and [1004](https://github.com/contao/core/blob/336c6a4e2ace2c878a797d23211140588b72b832/system/modules/core/classes/Theme.php#L1004). If someone adds a column with no `inputType` we can’t know which data it stores and therefore we shouldn’t modify the data of such columns, but just export it as it is. An extension developer has to convert special data himself via the `extractThemeFiles` and `exportTheme` hooks, as it is already the case for all other tables and value types, e.g. in [rocksolid-custom-elements/src/MadeYourDay/Contao/CustomElements.php:1271-1375](https://github.com/madeyourday/contao-rocksolid-custom-elements/blob/32041284353305cc162cbf3651448d680253a710/src/MadeYourDay/Contao/CustomElements.php#L1271-L1375). Why should we change that now for `tl_files`?

**Comment by ausi** [comment]
@leofeyer should we discuss this issue in the contao call next week?

**Comment by leofeyer** [comment]
Merged in c03b588dcee6ce23f7effa0ff5a5f77863b34a30.
